test/python/compiler/test_transpiler.py

```diff
@@ -1854,6 +1855,36 @@ def test_transpile_target_no_measurement_error(self, opt_level):
         res = transpile(qc, target=target, optimization_level=opt_level)
         self.assertEqual(qc, res)
 
+    def test_transpile_final_layout_updated_with_post_layout(self):
+        """Test that the final layout is correctly set when vf2postlayout runs.
+
+        Reproduce from #10457
+        """
+
+        def _get_index_layout(transpiled_circuit: QuantumCircuit, num_source_qubits: int):
+            """Return the index layout of a transpiled circuit"""
+            layout = transpiled_circuit.layout
+            if layout is None:
+                return list(range(num_source_qubits))
+
+            pos_to_virt = {v: k for k, v in layout.input_qubit_mapping.items()}
+            qubit_indices = []
+            for index in range(num_source_qubits):
+                qubit_idx = layout.initial_layout[pos_to_virt[index]]
+                if layout.final_layout is not None:
+                    qubit_idx = layout.final_layout[transpiled_circuit.qubits[qubit_idx]]
+                qubit_indices.append(qubit_idx)
+            return qubit_indices
+
+        backend = FakeVigo()
+        qubits = 3
+        qc = QuantumCircuit(qubits)
+        for i in range(5):
+            qc.cx(i % qubits, int(i + qubits / 2) % qubits)
+
+        tqc = transpile(qc, backend=backend, seed_transpiler=4242)
+        self.assertEqual([3, 2, 1], _get_index_layout(tqc, qubits))
```

Comment:
Would it help to have a mock assertion to ensure that `VF2PostLayout` has definitely run, so we can be sure to update the test if changes to the transpiler seed potentially cause it not to get run?

Reply:
Yeah, that's a good call I neglected to consider that this is dependent on sabre's output. I'll also add a comment inline about this for easier future debug if the test fails because of a seemingly unrelated transpiler change.

Reply:
I added a custom callback to verify the pass ran and set a new layout in https://github.com/Qiskit/qiskit-terra/pull/10466/commits/58c38a55be26235a18008c726581923457efecf8